src/core/experimental/memutils.d

```diff
@@ -0,0 +1,332 @@
+            }
+            // Compensate for the last (at most) 32 bytes.
+            store32i_sse(temp-0x10, xmm0);
+        }
+    }
+}
+
+void Dmemset_naive(void *dst, const ubyte val, size_t n)
+{
+    ubyte *d = cast(ubyte*)dst;
+    for (size_t i = 0; i != n; ++i)
+    {
+        d[i] = val;
+    }
+}
+
+// NOTE(stefanos):
+// Range-checking is not needed since the user never
+// pass an `n` (byte count) directly.
+
+// Copied from std.traits
+import core.internal.traits: Unqual;
+
+package template ModifyTypePreservingTQ(alias Modifier, T)
+{
+    static if (is(T U ==          immutable U)) alias ModifyTypePreservingTQ =          immutable Modifier!U;
+    else static if (is(T U == shared inout const U)) alias ModifyTypePreservingTQ = shared inout const Modifier!U;
+    else static if (is(T U == shared inout       U)) alias ModifyTypePreservingTQ = shared inout       Modifier!U;
+    else static if (is(T U == shared       const U)) alias ModifyTypePreservingTQ = shared       const Modifier!U;
+    else static if (is(T U == shared             U)) alias ModifyTypePreservingTQ = shared             Modifier!U;
+    else static if (is(T U ==        inout const U)) alias ModifyTypePreservingTQ =        inout const Modifier!U;
+    else static if (is(T U ==        inout       U)) alias ModifyTypePreservingTQ =              inout Modifier!U;
+    else static if (is(T U ==              const U)) alias ModifyTypePreservingTQ =              const Modifier!U;
+    else                                             alias ModifyTypePreservingTQ =                    Modifier!T;
+}
+
+template OriginalType(T)
+{
+    template Impl(T)
+    {
+        static if (is(T U == enum)) alias Impl = OriginalType!U;
+        else                        alias Impl =              T;
+    }
+
+    alias OriginalType = ModifyTypePreservingTQ!(Impl, T);
+}
+
+enum bool isAggregateType(T) = is(T == struct) || is(T == union) ||
+                               is(T == class) || is(T == interface);
+
+private template AliasThisTypeOf(T)
+if (isAggregateType!T)
+{
+    alias members = __traits(getAliasThis, T);
+
+    static if (members.length == 1)
+    {
+        alias AliasThisTypeOf = typeof(__traits(getMember, T.init, members[0]));
+    }
+    else
+        static assert(0, T.stringof~" does not have alias this type");
+}
+
+template DynamicArrayTypeOf(T)
+{
+    static if (is(AliasThisTypeOf!T AT) && !is(AT[] == AT))
+        alias X = DynamicArrayTypeOf!AT;
+    else
+        alias X = OriginalType!T;
+
+    static if (is(Unqual!X : E[], E) && !is(typeof({ enum n = X.length; })))
+    {
+        alias DynamicArrayTypeOf = X;
+    }
+    else
+        static assert(0, T.stringof~" is not a dynamic array");
+}
+
+enum bool isDynamicArray(T) = is(DynamicArrayTypeOf!T) && !isAggregateType!T;
+enum bool isStaticArray(T) = __traits(isStaticArray, T);
+enum bool isArray(T) = isStaticArray!T || isDynamicArray!T;
+
+void Dmemset(T)(ref T dst, const ubyte val)
```

Comment:
Needs proper documentation. Also why do we call we need the `D` prefix?
Also, this should be  before all of its tests and specialization, s.t. a reader has a consistent flow.

Reply:
Yes, I'll move them along with the other layout changes.
The `D` prefix is from when the project started. There's no specific reason for it, I'm just afraid that it will collide with the libc versions.

Reply:
> I'm just afraid that it will collide with the libc versions.

I don't think it will as the C `memset` accepts `void*` whereas this here accepts actual D types. Also, people generally don't import things from `core.stdc`, so in most cases both symbols won't be in the same scope.

Reply:
Good. I made the change but left the `D` prefix in a couple of places for either conceptual reasons or practical reasons (for example, there is a `Dmemset` that gets `void *`).
I'm open to further suggestions though.